### PR TITLE
[ADP-2945] remove encode decode addresses class

### PR DESCRIPTION
**Pull request by paolino**
- [x] rename ApiAddress to ApiAddressWithPath to free up the ApiAddress name
- [x] define a new `ApiAddress n` type in lieu of the tuple `(ApiT Address, Proxy n)`
- [x] remove special address encoding for the TypeSpec suit (overlapping hack is gone), by using valid encoded addressesfor the tests
- [x] remove `EncodeAddress n` and `DecodeAddress n` class and use a function over the `SNetworkId n` singleton.

ADP-2945

### CODE REVIEW

File: lib/wallet/api/http/Cardano/Wallet/Api/Types/Primitive.hs [hunk 1]
```diff
@@ -353,18 +347,6 @@ instance ToJSON (ApiT TxScriptValidity) where
...
-    parseJSON x = do
-        let proxy = Proxy @n
-        addr <- parseJSON x >>= eitherToParser
-            . bimap ShowFmt ApiT
-            . decodeAddress @n
-        return (addr, proxy)
-instance {-# OVERLAPS #-} EncodeAddress n => ToJSON (ApiT Address, Proxy n)
-  where
-    toJSON (addr, _) = toJSON . encodeAddress @n . getApiT $ addr
-
```
Comment (jonathanknowles) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types/Primitive.hs:367]: 🎉

File: lib/wallet/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 2]
```diff
@@ -1943,12 +1942,9 @@ shelleyDecodeStakeAddress serverNetwork txt = do
...
 
-instance EncodeAddress 'Mainnet where
-    encodeAddress = shelleyEncodeAddress SL.Mainnet
-
-instance EncodeAddress ('Testnet pm) where
-    -- https://github.com/cardano-foundation/CIPs/tree/master/CIP5
-    encodeAddress = shelleyEncodeAddress SL.Testnet
+encodeAddress :: SNetworkId n -> W.Address -> Text
+encodeAddress SMainnet = shelleyEncodeAddress SL.Mainnet
+encodeAddress (STestnet _) = shelleyEncodeAddress SL.Testnet
```
Comment (Unisay) [inline lib/wallet/src/Cardano/Wallet/Shelley/Compatibility.hs]: JFYI, my personal stylistic preference:

```suggestion
encodeAddress :: SNetworkId n -> W.Address -> Text
encodeAddress = \case
    SMainnet -> shelleyEncodeAddress SL.Mainnet
    STestnet _ -> shelleyEncodeAddress SL.Testnet
```

**Review by Unisay — APPROVED** [review]
Great effort, thanks!

**Pull request merged**